File: api/context.c
#include <stdio.h>
#include <time.h>
#include <sys/timeb.h>
#include <unistd.h>

#include "context.h"

FILE *context_log;

void ctx_set_log(const char *name) {
	context_log = fopen(name, "a");
}

void ctx_log_msg(const char *func, const char *text, const char *type) {
	if(context_log) {
		struct timeb tmb;
		char hostname[256];

		ftime(&tmb);
		gethostname(hostname, 256);

        fprintf(
			context_log,
			"%ld.%d %s %d %d %s %s %s\n",
            tmb.time, tmb.millitm,
            hostname,
			getpid(),
			0, //gettid(),
            type, func, text
        );
	}
}

File: api/context.php
<?php
define("CTX_BMARK", "BMARK");
define("CTX_START", "START");
define("CTX_ENDOK", "ENDOK");
define("CTX_ENDER", "ENDER");

$_context_log = null;

function ctx_set_log($name) {
	global $_context_log;
	if($name) {
		$_context_log = fopen($name, "a");
	}
	else {
		$_context_log = null;
	}
}

function ctx_log_msg($func, $text, $type) {
	global $_context_log;
	if($_context_log) {
        fprintf(
			$_context_log,
			"%f %s %d %d %s %s %s\n",
            microtime(true), # returning a float is 5.0+
            php_uname('n'),  # gethostname() is 5.3+
			posix_getpid(),
			posix_getpid(), //gettid(),
            $type, $func, $text
        );
	}
}

function __get_func() {
	$stack = debug_backtrace();
	if(count($stack) < 3) {
		return "top-level";
	}
	$p = $stack[2];
	return $p['function'];
}

function ctx_log_bmark($text) {ctx_log_msg(__get_func(), $text, "BMARK");}
function ctx_log_start($text) {ctx_log_msg(__get_func(), $text, "START");}
function ctx_log_endok($text) {ctx_log_msg(__get_func(), $text, "ENDOK");}
function ctx_log_ender($text) {ctx_log_msg(__get_func(), $text, "ENDER");}
?>
